### PR TITLE
Bach sample improvements

### PR DESCRIPTION
**Pull request by thijs-obj**
Changes:
* Don't require the `DataFrame` to be materialized when calling `get_sample()`
* Don't materialize in `get_unsampled()`. Previously we did that if `df._group_by` was set
* Also support sampling grouped DataFrames
* The `is_materialized` function is not used anymore now. But I do require it in another PR, so haven't removed it here.

I think somebody (hendrik?) asked me last time when I implemented #318 whether we couldn't skip the materialization requirement. I think I claimed then that that was not possible, no idea why :man_shrugging: I think this should work. Ran into this while working on the is_materialized stuff

### CODE REVIEW

File: bach/bach/sample.py [hunk 1]
```diff
@@ -49,7 +46,7 @@ def get_sample(df: DataFrame,
 
             sql = f'''
                 create temporary table tmp_table_name on commit drop as
-                ({df.view_sql()});
+                ({to_sql(df.base_node)});
```
Comment (thijs-obj) [inline bach/bach/sample.py:49]: main change 1/2

File: bach/bach/sample.py [hunk 2]
```diff
@@ -62,23 +59,13 @@ def get_sample(df: DataFrame,
...
     new_base_node = SampleSqlModel(table_name=table_name, previous=original_node)
-
-    return DataFrame.get_instance(
-        engine=df.engine,
-        base_node=new_base_node,
-        index_dtypes=df.index_dtypes,
-        dtypes=df.dtypes,
-        group_by=None
-    )
+    return df.copy_override_base_node(base_node=new_base_node)
```
Comment (thijs-obj) [inline bach/bach/sample.py:62]: main change 2/2

File: bach/bach/dataframe.py [hunk 3]
```diff
@@ -554,6 +554,28 @@ def copy_override(
 
         return self.__class__(**args, **kwargs)
 
+    def copy_override_base_node(self, base_node: SqlModel) -> 'DataFrame':
```
Comment (jansentom) [inline bach/bach/dataframe.py:557]: should this logic not be in `copy_override`? supporting base_node overrides there makes no sense

File: bach/bach/partitioning.py [hunk 4]
```diff
@@ -58,9 +61,12 @@ def __init__(self, group_by_columns: List[Series]):
         for col in group_by_columns:
             if not isinstance(col, Series):
                 raise ValueError(f'Unsupported argument type: {type(col)}')
-            if col.expression.is_constant and not isinstance(col, SeriesInt64):
-                # Dictated by PG
-                raise ValueError('Non-integer constant group by series not supported')
+            if col.expression.is_constant:
```
Comment (jansentom) [inline bach/bach/partitioning.py:64]: smart

**Review by jansentom — APPROVED** [review]

File: bach/bach/dataframe.py (hunk 3 above)
Comment (thijs-obj) [reply bach/bach/dataframe.py:557]: Maybe. Didn't go that route as `copy_override` is already quite a complicated function. And it has a lot of parameters that would interact with this logic. e.g. what to do if both `base_node` and `group_by` are set, or `index`, or `series`, or `index_dyptes` or `series_dtyptes`

**Pull request merged**